### PR TITLE
Do not invoke the supervisor twice for the same MapAsync failure

### LINKED ISSUE #23888: Supervision.Decider called twice when mapAsync throws exception in a Future
While writing tests around error handling in an Akka Stream, we observed that the `Supervision.Decider` we provide to a materializer can get called multiple times for a single exception in the flow. Specifically, we noticed this occur when an exception is thrown in a Future within a `mapAsync`. We were using the `Supervision.Decider` to manage statistics about how often exceptions are thrown in our flow.

Attached is an example test that demonstrates the behavior. We expect the counter to be incremented exactly once, but it is incremented twice instead.

It would be ideal if either this behavior was documented in an obvious location (I would suggest https://doc.akka.io/docs/akka/2.5/scala/stream/stream-error.html#errors-from-mapasync), or even better, if this issue was fixed, and it was safe to assume that a `Supervision.Decider` is only called once per exception thrown in a flow.

```
import akka.actor.ActorSystem
import akka.stream.scaladsl.{Flow, Sink, Source}
import akka.stream.{ActorMaterializer, ActorMaterializerSettings, Supervision}
import org.scalatest.{AsyncFreeSpec, Matchers}

import scala.concurrent.{ExecutionContext, Future}

class Test extends AsyncFreeSpec with Matchers {
  val decider: Supervision.Decider = {
    case exception: StringIndexOutOfBoundsException =>
      counter += 1
      println(s"Received an out of bounds exception : $exception")
      Supervision.Resume
  }

  implicit val system = ActorSystem("test")
  implicit val ec = ExecutionContext.Implicits.global
  implicit val materializer = ActorMaterializer(
    ActorMaterializerSettings(system).withSupervisionStrategy(decider)
  )

  var counter = 0

  val testFlow = Flow[String].mapAsync(1) { string =>
    Future { string.charAt(10) }
  }

  "should increment counter by only one" in {
    val preTestCounterValue = counter
    preTestCounterValue shouldBe 0

    val flatBuffersFlowTest =
      Source.single("shortStr")
        .via(testFlow)
        .runWith(Sink.seq)

    flatBuffersFlowTest.map { _ =>
      counter shouldBe (preTestCounterValue + 1) //Fails
    }
  }
}
```

Scala version: 2.11.8
Akka version: 2.5.6

### PR DESCRIPTION
**Pull request by johanandren**
Plus additional simplification and cleanup of the stage

Fixes  #23888

### CODE REVIEW

**Comment by johanandren** [comment]
I hope I didn't do too big refactoring but I really found it to be a mess that stage.

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 1]
```diff
@@ -1116,7 +1135,7 @@ private[stream] object Collect {
     }
   }
 
-  val NotYetThere = Failure(new Exception)
+  val NotYetThere = Failure(new Exception with NoStackTrace)
```
Comment (raboof) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: Without a stacktrace this is a very sparse exception... perhaps add a message to it?

**Review by raboof — APPROVED** [review]
Still pretty intricate, but the new naming does make it easier to follow

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala (hunk 1 above)
Comment (patriknw) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: and shouldn't it be `RuntimeException`?

Comment (johanandren) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: It's never thrown, just used as a marker

Comment (patriknw) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: ok, it's internal I see now

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 2]
```diff
@@ -1136,63 +1155,81 @@ private[stream] object Collect {
...
-      private def holderCompleted(holder: Holder[Out]) = handleFailureOrPushElem.apply(holder.elem)
-
-      val futureCB = getAsyncCallback[Holder[Out]](holderCompleted)
+      lazy val decider = inheritedAttributes.get[SupervisionStrategy].map(_.decider)
+        .getOrElse(Supervision.stoppingDecider)
+
+      private val futureCB = getAsyncCallback[Holder[Out]](holder ⇒
+        holder.elem match {
+          case Success(_) ⇒ pushNextIfPossible()
+          case Failure(ex) if NonFatal(ex) ⇒
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: `case Failure(NonFatal(ex))` =>

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 3]
```diff
@@ -1136,63 +1155,81 @@ private[stream] object Collect {
...
+        }
 
-      override def onPull(): Unit = pushOne()
+      private def pullIfNeeded(): Unit = {
+        if (buffer.used < parallelism && !hasBeenPulled(in)) tryPull(in)
+      }
 
       setHandlers(in, out, this)
+
+      override def toString = s"MapAsync.Logic(buffer=$buffer)"
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: toString of unknown things in the buffer?

File: akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala [hunk 4]
```diff
@@ -1102,7 +1103,25 @@ private[stream] object Collect {
...
+    // To support both fail-fast when the supervision directive is Stop
+    // and not calling the decider multiple times (#23888) we need to cache the decider result and re-use that
+    private var cachedSupervisionDirective: OptionVal[Supervision.Directive] = OptionVal.None
+
+    def supervisionDirectiveFor(decider: Supervision.Decider, ex: Throwable): Supervision.Directive = {
+      cachedSupervisionDirective match {
+        case OptionVal.Some(d) ⇒ d
+        case OptionVal.None ⇒
+          val d = decider(ex)
+          cachedSupervisionDirective = OptionVal.Some(d)
```
Comment (patriknw) [inline akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: what happens if it's first resume and then later another exception? won't that use the previous cached value for the new ex?

Comment (patriknw) [reply akka-stream/src/main/scala/akka/stream/impl/fusing/Ops.scala]: second look, new Holder is created for each element

**Review by patriknw — APPROVED** [review]
LGTM

**Pull request merged**